src/number_theory/modular_forms/slash_actions.lean

```diff
@@ -0,0 +1,122 @@
+/-
+Copyright (c) 2022 Chris Birkbeck. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Chris Birkbeck
+-/
+import analysis.complex.upper_half_plane.basic
+import linear_algebra.general_linear_group
+import linear_algebra.special_linear_group
+
+/-!
+# Slash actions
+
+This file defines a class of slash actions, which are families of right actions of a given group
+parametrized by some Type. This is modeled on the slash action of `GL_pos (fin 2) ℝ` on the space
+of modular forms.
+-/
+
+open complex upper_half_plane
+
+open_locale upper_half_plane
+
+local prefix `↑ₘ`:1024 := @coe _ (matrix (fin 2) (fin 2) _) _
+
+local notation `GL(` n `, ` R `)`⁺ := matrix.GL_pos (fin n) R
+
+local notation `SL(` n `, ` R `)` := matrix.special_linear_group (fin n) R
+
+/--A general version of the slash action of the space of modular forms.-/
+class slash_action (β : Type*) (G : Type*) (α : Type*) (γ : Type*) [group G] [ring α]
+  [has_smul γ α] :=
+(map : β → G → α → α)
+(mul_zero : ∀ (k : β) (g : G), map k g 0 = 0)
+(one_mul : ∀ (k : β) (a : α) , map k 1 a = a)
+(right_action : ∀ (k : β) (g h : G) (a : α), map k h (map k g a) = map k (g * h) a )
+(smul_action : ∀ (k : β) (g : G) (a : α) (z : γ), map k g (z • a) = z • (map k g a))
+(add_action : ∀ (k : β) (g : G) (a b : α), map k g (a + b) = map k g a + map k g b)
+
+/--Slash_action induced by a monoid homomorphism.-/
+def monoid_hom_slash_action { β : Type*} {G : Type*} {H : Type*} {α : Type*} {γ : Type*}
+  [group G] [ring α] [has_smul γ α] [group H] [slash_action β G α γ] (h : H →* G) :
+  slash_action β H α γ:=
+{ map := (λ k g a, slash_action.map γ k (h(g)) a),
+  mul_zero := by {intros k g, apply slash_action.mul_zero k (h g), },
+  one_mul := by {intros k a, simp only [map_one], apply slash_action.one_mul,},
+  right_action := by {simp only [map_mul], intros k g gg a, apply slash_action.right_action,},
+  smul_action := by {intros k g a z, apply slash_action.smul_action, },
+  add_action := by {intros k g a b, apply slash_action.add_action, }, }
+
+namespace modular_forms
+
+noncomputable theory
+
+/--The weight `k` action of `GL(2, ℝ)⁺` on functions `f : ℍ → ℂ`. -/
+def slash : ℤ → GL(2, ℝ)⁺ → (ℍ → ℂ) → (ℍ → ℂ) := λ k γ f,
+  (λ (x : ℍ), f (γ • x) * (((↑ₘ γ).det) : ℝ)^(k-1) * (upper_half_plane.denom γ x)^(-k))
```

Comment:
Why not
```suggestion
def slash (k : ℤ) (γ : GL(2, ℝ)⁺) (f : ℍ → ℂ) (x : ℍ) : ℂ :=
f (γ • x) * (((↑ₘ γ).det) : ℝ)^(k-1) * (upper_half_plane.denom γ x)^(-k)
```

Reply:
Can you make this modification? It seems easier to read.

I see that there are a couple of comments above that are not marked "resolved" but they may be so. Can you please click the button if this is already the case?

Reply:
Yeah I just was resolving the small issues that occured in the proofs with this. tbh I thought the original defn looked nicer but I'm happy to go with this if you both prefer it :)